### ISSUE TITLE
Create 1D Simulation Plugin

### ISSUE THREAD

**Issue by ktbolt** [SimVascular/SimVascular#437]
A plugin is needed to provide an interface to the 1D solver. 

Some preliminary tasks:

1. 3D model to 1D input file conversion code. This should use the center line extraction code inside of SV so we can go directly from a 3D model without the user generated paths and segmentations.

2. Grab the boundary condition information from the GUI so the BC’s (such as RCR) get written into the 1D solver input.

3. Clean up of the 1D solver source code for release.

4. Develop a GUI to run the above.

**Comment by ktbolt** [comment]
I have created a **1D Simulation** plugin using the **Simulation** plugin code as a template. In addition to the actual code that implements the plugin, adding a new plugin requires adding and modifying perhaps 10 different files (so far). I am still searching for the file that needs to be modified or added that allows me to create a **1D Simulation** plugin from the **SV Data Manager**. 
<br>
<img width="1026" alt="Screen Shot 2019-03-24 at 9 35 40 PM" src="https://user-images.githubusercontent.com/23109458/54895452-de0fcf00-4e7c-11e9-8bad-be456c8c3887.png">

**Comment by ktbolt** [comment]
I've found the problem, missed changing a string name for a MITK data node

    mitk::NodePredicateDataType::New("sv4guiSimulation1dFolder")

Note that the string **sv4guiSimulation1dFolder** is used in nine different places in code outside of the **1D Simulation** plugin. We need place all the information needed to create a new plugin in one location and encapsulate the data needed to define a plugin it in a class.

**Comment by ktbolt** [comment]
Spent some time learning how to use **Qt Creator** to add a **1D Mesh** toolbox page to the **SV 1D Simulation** plugin to handle reading a mesh, generating a mesh, etc., nice way to group related widgets.
<br>
![1dsim](https://user-images.githubusercontent.com/23109458/55191740-31756c00-5160-11e9-97ec-ff63a11de2b2.png)

**Comment by ktbolt** [comment]
I spent a good part of the day trying to figure out why opening a saved project with a **SV 1D Simulation** plugin defined in it would bring up an **SV Simulation** plugin instead. 

I turns out that SV determines the plugin type from the **file extension** of the data node files stored in a  plugin's directory. I was using the same file extension (**.sjb**) as the **SV Simulation** plugin so SV thought that the **SV 1D Simulation** plugin was an **SV Simulation** plugin. I changed the **SV 1D Simulation** file extension to **.s1djb** and everything now works.

**Comment by ktbolt** [comment]
I've been restructuring the **generate_1Dinput_from_3Dcenterlines.py** script, adding argument processing, modularizing the code, ...

There are lots of problems getting vmtk to work on a Mac, current vmtk version is installed using conda but there still are problems with certain Python versions it seems. I can build vmtk from source fine but it will not execute correctly.

**Comment by ktbolt** [comment]
I got VMTK 1.4 centerlines working on MacOS, had to modify the VMTK source to work correctly with Python 3.6.

**Comment by ktbolt** [comment]
I've mostly finished with restructuring **generate_1Dinput_from_3Dcenterlines.py**, a bit more cleanup to do. I broke up the original 800 line script into several classes
```
    centerlines.py
    generate-1d-mesh.py
    manage.py
    mesh.py
    parameters.py
    utils.py
```
to better mange the different ways it can be used.

**Comment by rohanshad** [comment]
Is this something that would find use in refining boundary conditions for 3D simulations? If so, would there be a way to write a script that for example, would iterate through different combinations of RCR outlet conditions until a target Pressure is reached?

**Comment by alisonmarsden** [comment]
We have a number of strategies for automated tuning in the lab, and you’re welcome to come talk to us about them.   They may not be in the open source release yet since we’re still working on them.    Using the 1D solver for this in the future would certainly be a possibility.

**Comment by ktbolt** [comment]
I am now able to execute the **generate-1d-mesh** Python script to generate a 1D mesh from the **1D Simulation** plugin, display the centerlines (green) and the 1D mesh (red). 
<br>
<img width="1387" alt="Screen Shot 2019-04-11 at 8 12 03 PM" src="https://user-images.githubusercontent.com/23109458/56009926-6fad7800-5c96-11e9-94e5-4d2905600ee7.png">
<br>
I will use the vmtk c++ code in SV to compute centerlines because SV does not include the vmtk Python package. I will keep the centerline computation in the Python script so it can be used standalone if needed.

**Comment by wgyang** [comment]
Great!

Sent from my iPhone

On Apr 11, 2019, at 8:23 PM, Dave Parker <notifications@github.com<mailto:notifications@github.com>> wrote:


I am now able to execute the generate-1d-mesh Python script to generate a 1D mesh from the 1D Simulation plugin, display the centerlines (green) and the 1D mesh (red).


[Screen Shot 2019-04-11 at 8 12 03 PM]<https://user-images.githubusercontent.com/23109458/56009926-6fad7800-5c96-11e9-94e5-4d2905600ee7.png>


I will use the vmtk c++ code in SV to compute centerlines because SV does not include the vmtk Python package. I will keep the centerline computation in the Python script so it can be used standalone if needed.

**Comment by ktbolt** [comment]
I've added creating the 1D solver input files from the plugin. 

Next I'll integrate execution of the 1D solver.

**Comment by ktbolt** [comment]
I've added code to look for the 1D solver and execute it with the solver input file, seems to work!

**Comment by ktbolt** [comment]
The differences in simulation results was from the inlet / outlet area calculation. The original script used  3.14 in the area calculation which I replaced with the NumPy math pi constant (np.pi) in the new script. Using np.pi in the old script produces simulation results that match exactly.

**Comment by ktbolt** [comment]
I've changed the way I'm calling the Python 1d mesh generation script to use the Python PyObject_CallMethod() instead of PyRun_SimpleString() so I can return exception messages back to SV for display.

The 1d mesh generation script now writes a log file that is displayed in SV if anything goes wrong.

Everything seems to be working. I'll demo for Weiguang and Melody to see what they think.

**Comment by ktbolt** [comment]
I've modified the CMake files to build installers (Ubuntu and MacOS) for SV with the 1D solver plugin. It's a bit of a hack the way I am installing the Python scripts, will figure out a better strategy for this later.

@melodydong I would like add some more checks to the plugin and do a bit more testing before giving the installers to you and @wgyang to test.

**Comment by ktbolt** [comment]
Using char*, std:string and QString in SV and their inoperability causes very much pain. We need to use std::string in all non-GUI code and restrict QString to just GUI code.

**Comment by ktbolt** [comment]
I've added code to check for valid BCs and activate push buttons to generate solver files when all BCs and solver parameters have been set.

I also added a check to notify the user if model geometry has changed. MITK does not appear to be able to check if a Data Node has modified data using a time stamp or the like, so I had to implement this using properties.

**Comment by ktbolt** [comment]
I've redone the way BC data is written and processed in the 1D plugin and added code to the Python script to incorporate resistance boundary conditions, didn't know (or forgot!) we needed this for the first release.

**Comment by ktbolt** [comment]
I've fixed a problem with returning information when an exception occurs. I've also modified the code to read in inflow bc data to accept a space or a comma as a delimiter.

**Comment by melodydong** [comment]
Students with Windows computers are having issues generating the solver.in file.  There is no error generated and all of the other files, except for the solver.in, are saved into the Simulations1d folder. Projects where this occurred are on Skalak under /home/SHARED/SimVascular/Class_proj.  The Luengo_Midterm project, even after fixing the .flow file will not generate a solver.in on their Windows machine.

**Comment by osmsc** [comment]
Hi Melody - is the problem with Leungo_Midterm platform specific now?  Originally the problem was also on Linux.  Does it now work on Linux and not Linux?

**Comment by melodydong** [comment]
Hi Nathan,  It seems that it is platform specific now.  The problem occurred on all platforms before because of the comma-delimited flow file.  Now with a fixed comma-delimited flow file, a solver.in can be generated on Linux, but not on Windows. This also seems to be a similar issue to my original AortaFem_Demo project that would not generate a solver.in on Windows, even on your Windows machine.

**Comment by ktbolt** [comment]
@melodydong The **Aorta_fem** project you placed on skalak looks odd:

`Images/		Models/		Segmentations/	Simulations1d/	ao_fem.s1djb
Meshes/		Paths/		Simulations/	ao_fem/
`

The **ao_fem** directory and **ao_fem.s1djb** file should be under the **Simulations1d** directory.

**Comment by melodydong** [comment]
@ktbolt That must have been an old version.  The AortaFem1D_Demo.zip has the folders in the correct place, but still does not generate a solver.in with Windows. Another student's project (Post-OpIntracranianAneurysmProject.zip in the same folder) is another example of a project that generates a solver.in on Linux but not on Windows.

**Comment by osmsc** [comment]
The problem with Leungo_Midterm "final3.s1djb" is the "Flow Rate" file still.  In Notepad, you need to replace all occurrences of:

&#x0D;

with an empty string.

**Comment by ktbolt** [comment]
@osmsc I am testing on Windows using Melody's **AortaFem1D_Demo** project. The flow file data is being stored correctly into the **aofem_dmo.s1djb** file but when it writes it out again an extra crtl-M is written at the end of the first line:

$ od -c steadyflow.txt
   0 . 0   65 \r\r\n  1 . 0   6 5

The extra \r is causing the problem. 

I don't see where in the code this could be occurring. I will modify the Python script to deal with this.

**Comment by ktbolt** [comment]
Several users are seeing centerlines geometry with incorrect connectivity.

![cl-problem](https://user-images.githubusercontent.com/23109458/57492439-a9fa3d00-7275-11e9-9928-baab82ea37f5.png)

Note that calculating centerlines for this model from the Modeling Tool crashes SV. 

I believe that the vmtk centerlines calculation creates a Delaunay triangulation so I am wondering if this problem is caused by a bad (non-manifold) surface mesh. The model shown here was not able to mesh.

**Comment by ktbolt** [comment]
The  **vmtkcenterlines** script creates correct centerlines so the problem must be what data (e.g. seed point IDs) is being passed to the c++ **vtkvmtkPolyDataCenterlines** function.

**Comment by osmsc** [comment]
@ktbolt - Dave, can you convert the python scripts for the 1-D plugin into a package that can be included in "Python/site-packages", something like:

site-packages/sv_1d_simulation

that can be imported into Python via:

import sv_1d_simulation

and called  from the C++.  Currently all of the scripts are in the "toplevel" of the directory which isn't modular.

**Comment by ktbolt** [comment]
@osmsc The Python script will be placed in a separate GitHub repository so people can use it stand alone. During SV build the Python code will be moved into SV Python/site-packages. I'm not sure what to rename the package to, maybe something like sv_1d_simulation_pre?

**Comment by ktbolt** [comment]
I have fixed the problem with centerlines. In the **sys_geom_centerlines** function cap IDs were being copied using

`  for (pointId = 0;pointId<nsources;pointId++)
  {
    capInletIds->InsertNextId(*sources+pointId);
  }
  for (pointId = 0;pointId<ntargets;pointId++)
  {
    capOutletIds->InsertNextId(*targets+pointId);
  }
`
Accessing arrays like this is a really bad practice and can lead (obviously) to buggy code, how centerlines ever worked is unknown.  I replaced the above code with

`  for (pointId = 0;pointId<nsources;pointId++)
  {
    capInletIds->InsertNextId(sources[pointId]);
  }
  for (pointId = 0;pointId<ntargets;pointId++)
  {
    capOutletIds->InsertNextId(targets[pointId]);
  }
`
Centerlines now work

<img width="689" alt="Screen Shot 2019-05-12 at 9 13 28 PM" src="https://user-images.githubusercontent.com/23109458/57595997-20f53700-74fd-11e9-9745-de8aeb6d64c3.png">

The SV geometry code needs to be cleaned up, create a better interface (some arrays are copied three times between function calls), and use c++ best practices (e.g. get rid of pointers and use std::vector).

**Comment by melodydong** [comment]
![image](https://user-images.githubusercontent.com/16911253/57644245-8df5e480-7570-11e9-8985-62d3fbe1f5bd.png)

![image](https://user-images.githubusercontent.com/16911253/57644291-acf47680-7570-11e9-964f-0f988ecad47c.png)


Users have been getting blank errors when running a 1D simulation.  No .log files are outputted and no files are generated.  This might be due to the 1D Solver itself not being able to complete, due to extreme tapering, poor boundary conditions, or other errors with the solver.in that makes the 1D solver fail.

**Comment by ktbolt** [comment]
@melodydong When I run the 1D solver with assertions (these check for error conditions but are not enabled for the release solver) enabled I get

At PA2Group4_Seg2 The area goes to -0.0578512

In the solver.in file I see:

SEGMENT Group4_Seg2 2 10.78432102184993 108 2 3 2.680661339856945 0.07566014018744738 0.0 MAT1 NONE 0.0 0 0 RESISTANCE RESISTANCE_0

The outlet area **0.07566014018744738** for this segment is quite small so I guess it is going negative during the simulation.

When I run the 1D simulation on Ubuntu from the GUI it finishes without an error but sv1dsolver.log file contains NaNs. What OS are you running on?

The 1D solver really needs to throw an exception if it is calculating NaNs!

**Comment by ktbolt** [comment]
I've made changes to the 1D solver to throw exceptions for NaN residuals and negative areas.

**Comment by melodydong** [comment]
@ktbolt @osmsc Below are a list of issues that we've found while students have been using the 1D solver in class.  All of the issues below should be reproduceable with any of the student's projects that are on Skalak (/home/SHARED/SimVascular/Class_projects/)


1.  If there are multiple models & multiple simulations, when a user creates a second simulation, it will automatically use the model from the 1st simulation.
2. When you delete a 1D simulation job, it will not delete from the original project directory. Users may then have the issues described in the first bullet point.
3. Need a better way to import old projects into the new SimVascular without changing .svproj and creating a Simulation1d folder.  This caused a lot of issues with students/users doing this incorrectly.
4. On Windows, there is a resolution issue where you cannot fully see the text in the boxes of the Simulation1d GUI plugin.
5. Simulation1d Basic Parameters Box: if you change any of these values, it does not update the solver.in, specifically the density and viscosity should be updated in the MATERIAL CARD of the solver.in.  The IC file and initial pressure should be greyed out/not an option for the 1D simulation file (not implemented in the 1D solver yet...)
6. Need more documentation on which way flow is supposed to be prescribed.  For the 3D solver, convention is flow to be negative when going into the model.  For 1D solver, it is the opposite.  This may cause some confusion for users.
7. Solver.in and SimVascular in general is very sensitive to names with spaces in them.  Either have this is as a check for users to not name paths/segmentations/models with spaces in it, or flash an error/warning when they do to make sure they change it.
8. Wall properties: deformable wall options do not work, rigid option defaults to a hard-coded material property (probably specified in the generate_centerlines.py script)
9. For wall properties, need option to specify an Olufsen or Linear material model and option for k1, k2, k3 values and exponent for Olufsen or Eh/r value for Linear as well as a reference pressure for both. (as a note, the 1D solver only does a deformable wall, but you can model "rigid" walls by increasing the elastic modulus for the material properties.
10. For deformable variable, must be able to choose each specific segment and material properties.
11. solver.in comments need to be updated & cleaned up with only boundary conditions that the 1D solver supports currently (generate_centerlines.py might be using an older version of solver.in comments).
12. (enhancement) Have an automatic simulation filename = model name
13. Change group name ID's to vessel names that the user has specified - could be for post-processing.
14. Check if flow file (time points) are out of order, correct them and give user a warning.  One student had a flow file in an incorrect order and had trouble debugging why.

**Comment by ktbolt** [comment]
@melodydong Thanks for the list of issues! Only 14? Not so bad ...

Data Nodes with names containing spaces (7) is a big problem, this just came up in the SV Forum where svsolver was failing because of a file name with spaces. I will file a bug for this.

Several of these issues are things I've just not implemented yet (e.g. 5, 8, 9, 10), should not be that hard to add them.

Issues 1 and 2 I've not seen, will have a look.

There are other enhancements, like setting inlet / outlet BCs from a Simulation Tool or read them from a file, that I will be adding.

**Comment by ktbolt** [comment]
I have started looking into 

`1) If there are multiple models & multiple simulations, when a user creates a second simulation, it will automatically use the model from the 1st simulation.`

The 1D Simulation Tool GUI panel is not being updated with the values set for different instances of the tool. It turns out that the **CreateQtPartControl** method (defined for all tools) that sets up GUI widgets is called only once for all instances, so the widgets are share between all instances.

**Comment by ktbolt** [comment]
I have fixed issue 1), figured out (mostly) the interaction between the gui and job objects, cleaned up the code.

**Comment by ktbolt** [comment]
I'm now looking into 

`2) When you delete a 1D simulation job, it will not delete from the original project directory. Users may then have the issues described in the first bullet point.`

SimVascular is not consistent in the way it removes the directory / files created by a plugin when a Data Node is removed. The directories for Simulation tools are not removed but their jobs (.sjb) files are. The files created by Mesh tools are not removed unless you save the project.

**Comment by ktbolt** [comment]
I have found where SV is saving and deleting files (**sv4guiProjectManager::SaveProject**) and fixed the problem with the plugin not deleting job files, missed adding a **1d** to a variable. It would be good to separate saving and deleting project files into two functions I think, the **sv4guiProjectManager::SaveProject** is almost 400 lines, easy to mess things up in that jungle of variables. 

I've also been looking into deleting job directories, think I found the Qt functions to do that. 

Project files don't actually get removed when you delete a Data Node. That doesn't happen until you manually save the project. This could lead to user confusion (it was confusing me for a good while) ,delete a Data Node and not save the project and the Node appears the next time the project is opened.

**Comment by melodydong** [comment]
On Windows, Simvascular with the 1D plugin does not link to the 3D solver and I am not able to run a 3D simulation.  I've tried linking the corresponding presolver, svsolver, and postsolver under Windows->SimVascular Simulation.  It will generate simulation files, but will not execute the simulation without MPI (even though I have unchecked MPI).

**Comment by ktbolt** [comment]
@melodydong This problem is not related to the 1D Solver plugin. The latest Windows installer also included modifications to the Simulation Preferences code that identifies the mpi implementation and svsolver binary locations but is not working correctly on Windows.

@osmsc How can we fix this?

**Comment by ktbolt** [comment]
I just discovered a few things about how SV works, like how there is only a single object created for all plugin instances, that will require me to re-implement some of the plugin code that deals with displaying geometry and storing plugin state.

This will take me at least a couple of days, first to better understand how SV handles geometry and then to re-do the plugin.

**Comment by osmsc** [comment]
It looks like I accidentally forgot to include the nompi version of svsolver with the latest release (2019-01-25) on Windows.  The mpi version should run if you specify mpi, presolver, etc.  I need to fix the registry lookup for the automatic detection.  Will do this by Monday.

**Comment by ktbolt** [comment]
The first version of this is done and working ok although there are still problems when multiple instances are created. I will open a separate issue for that.
